src/query/storage/promremote/storage.go

```
@@ -0,0 +1,158 @@
+	encoded, err := encodeWriteQuery(query)
+	if err != nil {
+		return err
+	}
+
+	var wg sync.WaitGroup
+	multiErr := xerrors.NewMultiError()
+	var errLock sync.Mutex
+	for _, endpoint := range p.opts.endpoints {
+		endpoint := endpoint
+		if endpoint.resolution == query.Attributes().Resolution &&
+			endpoint.retention == query.Attributes().Retention {
+			metrics := p.endpointMetrics[endpoint.name]
+			wg.Add(1)
+			go func() {
+				defer wg.Done()
+				err := p.writeSingle(ctx, metrics, endpoint.address, bytes.NewBuffer(encoded))
+				if err != nil {
+					errLock.Lock()
+					multiErr = multiErr.Add(err)
+					errLock.Unlock()
+					return
+				}
+			}()
+		}
+	}
+
+	wg.Wait()
+
+	if !multiErr.Empty() {
+		return multiErr
+	}
+	return nil
+}
+
+func (p *promStorage) Type() storage.Type {
+	return storage.TypeRemoteDC
+}
+
+func (p *promStorage) Close() error {
+	p.client.CloseIdleConnections()
+	return nil
+}
+
+func (p *promStorage) ErrorBehavior() storage.ErrorBehavior {
+	return storage.BehaviorFail
+}
+
+func (p *promStorage) Name() string {
+	return "prom-remote"
+}
+
+func (p *promStorage) writeSingle(
+	ctx context.Context,
+	metrics *instrument.MethodMetrics,
+	address string,
+	encoded io.Reader,
+) error {
+	req, err := http.NewRequestWithContext(ctx, http.MethodPost, address, encoded)
+	if err != nil {
+		return err
+	}
+	req.Header.Set("content-encoding", "snappy")
+	req.Header.Set(xhttp.HeaderContentType, xhttp.ContentTypeProtobuf)
+
+	start := time.Now()
+	resp, err := p.client.Do(req)
+	methodDuration := time.Since(start)
+	if err != nil {
+		metrics.ReportError(methodDuration)
+		return err
+	}
+	metrics.ReportSuccess(methodDuration)
+	defer func() { _ = resp.Body.Close() }()
+
+	if resp.StatusCode/100 != 2 {
+		metrics.ReportError(methodDuration)
+		response, err := ioutil.ReadAll(resp.Body)
+		if err != nil {
+			response = []byte(fmt.Sprintf("error reading body: %v", err))
```

Not sure if we want to expose those errors to the client.

---

I will add TODO and will fix in another PR where I have logger available.

---

So you can return a generic error now, and TODO just logging.